imap_processing/ultra/l0/decom_ultra.py

```diff
@@ -1,269 +1,226 @@
+            all_events.append(empty_event)
+            all_indices.append(i)
         else:
-            for i in range(count):
-                logging.info(f"Appending image #{i}")
-                append_params(decom_data, packet)
-
-    return decom_data
+            # Here there are multiple images in a single packet,
+            # so we need to loop through each image and decompress it.
+            event_data_list = read_image_raw_events_binary(eventdata_array[i], count)
+            all_events.extend(event_data_list)
+            # Keep track of how many times does the event occurred at this epoch.
+            all_indices.extend([i] * count)
+
+    # Now we have the event data, we need to create the xarray dataset.
+    # We cannot append to the existing dataset (sorted_packets)
+    # because there are multiple events for each epoch.
+    idx = np.array(all_indices)
+
+    # Expand the existing dataset so that it is the same length as the event data.
+    expanded_data = {
+        var: ds[var].values[idx] for var in ds.data_vars if var != "eventdata"
+    }
 
+    # Add the event data to the expanded dataset.
+    for key in event_data_list[0]:
+        expanded_data[key] = np.array([event[key] for event in all_events])
 
-def process_ultra_aux(sorted_packets: list, decom_data: dict) -> dict:
-    """
-    Unpack and decode Ultra AUX packets.
+    event_ids = get_event_id(expanded_data["shcoarse"])
 
-    Parameters
-    ----------
-    sorted_packets : list
-        AUX packets sorted by time.
-    decom_data : collections.defaultdict
-        Empty dictionary.
+    coords = {
+        "epoch": ds["epoch"].values[idx],
+        "event_id": ("epoch", event_ids),
+    }
 
-    Returns
-    -------
-    decom_data : dict
-        A dictionary containing the decoded data.
-    """
-    for packet in sorted_packets:
-        append_params(decom_data, packet)
+    dataset = xr.Dataset(coords=coords)
+    for key, data in expanded_data.items():
+        dataset[key] = xr.DataArray(
+            data,
+            dims=["epoch"],
+        )
 
-    return decom_data
+    return dataset
 
 
-def process_ultra_rates(sorted_packets: list, decom_data: dict) -> dict:
+def process_ultra_rates(ds: xr.Dataset) -> xr.Dataset:
     """
     Unpack and decode Ultra RATES packets.
 
     Parameters
     ----------
-    sorted_packets : list
-        RATES packets sorted by time.
-    decom_data : collections.defaultdict
-        Empty dictionary.
+    ds : xarray.Dataset
+       Rates dataset.
 
     Returns
     -------
-    decom_data : dict
-        A dictionary containing the decoded data.
+    dataset : xarray.Dataset
+        Dataset containing the decoded and decompressed data.
     """
+    decom_data = defaultdict(list)
     if (
         isinstance(ULTRA_RATES.mantissa_bit_length, int)
         and isinstance(ULTRA_RATES.len_array, int)
         and isinstance(ULTRA_RATES.block, int)
         and isinstance(ULTRA_RATES.width, int)
     ):
```

Comment:
if this isn't true do we want to just skip this, or should we raise / log that happening here? It seems like these statements should always be true unless something bad is happening.